port/cpl_http.cpp

```diff
@@ -553,41 +554,65 @@ constexpr TupleEnvVarOptionName asAssocEnvVarOptionName[] = {
-                : CPLGetConfigOption(asAssocEnvVarOptionName[i].pszEnvVar,
-                                     nullptr);
+            pszFilename ? VSIGetPathSpecificOption(pszFilename,
+                                                   sTuple.pszEnvVar, nullptr)
+                        : CPLGetConfigOption(sTuple.pszEnvVar, nullptr);
         if (pszVal != nullptr)
         {
-            papszOptions = CSLSetNameValue(
-                papszOptions, asAssocEnvVarOptionName[i].pszOptionName, pszVal);
+            aosOptions.AddNameValue(sTuple.pszOptionName, pszVal);
         }
     }
-    return papszOptions;
+    return aosOptions.StealList();
 }
 
 /************************************************************************/
 /*                      CPLHTTPGetNewRetryDelay()                       */
 /************************************************************************/
 
-double CPLHTTPGetNewRetryDelay(int response_code, double dfOldDelay,
-                               const char *pszErrBuf, const char *pszCurlError)
+/** Return the new retry delay.
+ *
+ * This takes into account the HTTP response code, the previous delay, the
+ * HTTP payload error message, the Curl error message and a potential list of
+ * retriable HTTP codes.
+ *
+ * @param response_code HTTP response code (e.g. 400)
+ * @param dfOldDelay Previous delay (nominally in second)
+ * @param pszErrBuf HTTP response body of the failed request (may be NULL)
+ * @param pszCurlError Curl error as returned by CURLOPT_ERRORBUFFER (may be NULL)
+ * @param pszRetriableCodes nullptr to limit to the default hard-coded scenarios,
+ * "ALL" to ask to retry for all non-200 codes, or a comma-separated list of
+ * HTTP codes (e.g. "400,500") that are accepted for retry.
+ * @return the new delay, or 0 if no retry should be attempted.
+ */
+static double CPLHTTPGetNewRetryDelay(int response_code, double dfOldDelay,
+                                      const char *pszErrBuf,
+                                      const char *pszCurlError,
+                                      const char *pszRetriableCodes)
 {
-    if (response_code == 429 || response_code == 500 ||
-        (response_code >= 502 && response_code <= 504) ||
-        // S3 sends some client timeout errors as 400 Client Error
-        (response_code == 400 && pszErrBuf &&
-         strstr(pszErrBuf, "RequestTimeout")) ||
-        (pszCurlError && (strstr(pszCurlError, "Connection timed out") ||
-                          strstr(pszCurlError, "Operation timed out") ||
-                          strstr(pszCurlError, "Connection reset by peer") ||
-                          strstr(pszCurlError, "Connection was reset"))))
+    bool bRetry = false;
+    if (pszRetriableCodes && pszRetriableCodes[0])
+    {
+        bRetry = EQUAL(pszRetriableCodes, "ALL") ||
+                 strstr(pszRetriableCodes, CPLSPrintf("%d", response_code));
+    }
+    else if (response_code == 429 || response_code == 500 ||
```

Comment:
Probably out of scope for this PR but a nice readability improvement would be to replace these literal error codes with constants from a header (I suppose CURL has one, or something like https://github.com/j-ulrich/http-status-codes-cpp/tree/main).